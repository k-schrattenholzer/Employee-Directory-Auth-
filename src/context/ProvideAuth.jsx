import { createContext, useContext, useMemo, useState } from 'react';
import { getUser } from '../services/users.js';

const ProvideAuth = createContext();

const UserProvider = ({ children }) => {
  const currentUser = getUser();
  const [user, setUser] = useState(
    currentUser ? { id: currentUser.id, email: currentUser.email } : {}
  );

  const value = useMemo(() => ({ user, setUser }), [user]);

  return <ProvideAuth.Provider value={value}>{children}</ProvideAuth.Provider>;
};

const useUser = () => {
  const context = useContext(ProvideAuth);

  if (context === undefined) {
    throw new Error('useUser must be used within a UserProvider');
  }

  return context;
};

export { ProvideAuth, UserProvider, useUser };
