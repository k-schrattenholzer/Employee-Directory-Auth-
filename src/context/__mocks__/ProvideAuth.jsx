import { createContext, useContext, useMemo, useState } from 'react';


const ProvideAuth = createContext();

const UserProvider = ({ mockUser, children }) => {

  const [user, setUser] = useState(
    mockUser ? { id: mockUser.id, email: mockUser.email } : {}
  );

  const value = useMemo(() => ({ user, setUser }), [user]);

  return <ProvideAuth.Provider value={value}>{children}</ProvideAuth.Provider>;
};

const useUser = () => {
  const context = useContext(ProvideAuth);

  if (context === undefined) {
    throw new Error('useUser must be used within a UserProvider');
  }

  return context;
};

export { UserProvider, useUser }
