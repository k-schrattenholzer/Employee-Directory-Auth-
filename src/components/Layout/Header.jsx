import { useHistory } from "react-router-dom";
import { useUser } from "../../context/ProvideAuth.jsx"

import { signOutUser } from "../../services/users.js";

export default function Header() {

  const { user, setUser } = useUser();
  const history = useHistory();

  const handleSignOut = async () => {
    await signOutUser();
    setUser({});
    history.push('/login')
  }

  return (
    <div>
      <header>
        <p>{user?.email ? `Signed in as ${user?.email}` : 'Not Signed In' }</p>
        {user?.email ? <button onClick={handleSignOut}>Sign Out</button> : '' }

      </header>
    </div>
  )
}
