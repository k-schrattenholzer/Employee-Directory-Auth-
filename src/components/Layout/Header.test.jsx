import { render, screen } from '@testing-library/react';
import { MemoryRouter } from 'react-router-dom';
import { UserProvider } from '../../context/ProvideAuth.jsx';
import Header from './Header.jsx';

jest.mock('../../context/ProvideAuth.jsx');
jest.mock('../../services/users.js');

it('should display a header with `Not Signed In` when signed out', () => {
  const { container } = render(
    <UserProvider>
      <MemoryRouter>
        <Header />
      </MemoryRouter>
    </UserProvider>
  );
  expect(container).toMatchSnapshot();
});

it('should display a header with users email displayed when signed in', () => {
  const { container } = render (
    <UserProvider mockUser={{ id: 1, email: 'test@example.com'}}>
      <MemoryRouter>
        <Header/>
      </MemoryRouter>
    </UserProvider>
  )
  screen.getByText('test@example.com', { exact: false });
  expect(container).toMatchSnapshot();
});
