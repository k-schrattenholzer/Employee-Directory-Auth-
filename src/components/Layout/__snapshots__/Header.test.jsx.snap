// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should display a header with \`Not Signed In\` when signed out 1`] = `
<div>
  <div>
    <header>
      <p>
        Not Signed In
      </p>
      
    </header>
  </div>
</div>
`;

exports[`should display a header with users email displayed when signed in 1`] = `
<div>
  <div>
    <header>
      <p>
        Signed in as test@example.com
      </p>
      <button>
        Sign Out
      </button>
    </header>
  </div>
</div>
`;
